doc/specs/stdlib_string_type.md

```diff
@@ -0,0 +1,1607 @@
+---
+title: string type
+---
+
+# The `stdlib_string_type` module
+
+[TOC]
+
+## Introduction
+
+The `stdlib_string_type` provides a derived type holding an arbitrary sequence
+of characters compatible with most Fortran intrinsic character procedures as
+well as operators for working with character variables and constants.
+
+
+## Derived types provided
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### The `string_type` derived type
+
+The `string_type` is defined as a non-extenable derived type representing a
+sequence of characters. The internal representation of the character sequence
+is implementation dependent and not visible for the user of the module.
+
+#### Status
+
+Experimental
+
+
+## Procedures and methods provided
+
+Procedures returning `string_type` instances can usually used in elemental
+context, while procedures returning scalar character values can only be
+used in a pure way.
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Constructor for empty string
+
+The module defines a default constructor to create an empty string type.
+
+#### Description
+
+Creates a string instance representing an empty string.
+
+#### Syntax
+
+`res = [[stdlib_string_type(module):string_type(interface)]] ()`
+
+#### Status
+
+Experimental
+
+#### Class
+
+Elemental function.
+
+#### Argument
+
+None.
+
+#### Result value
+
+The result is an instance of `string_type`.
+
+#### Example
+
+```fortran
+use stdlib_string_type
+implicit none
+type(string_type) :: string
+string = string_type()
+! len(string) == 0
+end
+```
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Constructor from character scalar
+
+The module defines a default constructor to create a string type
+from a character scalar.
```

Comment:
Just a question: since the internal representation is private, the default structure constructor would not work correctly?

Reply:
Default constructors for allocatables have been buggy for several compilers in my experience. Best to avoid those whenever possible.

Reply:
I feel inclined to drop the word "default" in the first sentence. Given that the constructors over-ride the *default* structure constructor, I think it makes the sentence more confusing. I suppose the meaning of default above was in the sense of "most commonly used" constructor.